### PR TITLE
kubeadm: init must validate or generate token before anything else.

### LINKED ISSUE #112: kubeadm init with token should validate token earlier
Currently, `kubeadm init --discovery token://...` validates the format of the passed token only after a few steps are executed, i.e. certificate generation, if none exist. If the token is invalid, one needs to run `kubeadm reset` before retrying.

Therefore, when a token is passed, a token format validation should happen earlier.

### PR DESCRIPTION
**Pull request by pires**
**What this PR does / why we need it**: `kubeadm init` must validate or generate a token before anything else. Otherwise, if token validation or generation fail, one will need to run `kubeadm reset && systemctl restart kubelet` before re-running `kubeadm init`.

**Which issue this PR fixes**: fixes kubernetes/kubeadm#112

**Special notes for your reviewer**: /cc @luxas

Tested manually.

### With no token

```
$ sudo ./kubeadm init
[kubeadm] WARNING: kubeadm is in alpha, please do not use it for production clusters.
[preflight] Running pre-flight checks
[init] Using Kubernetes version: v1.5.2
[token-discovery] A token has not been provided, generating one
[certificates] Generated Certificate Authority key and certificate.
[certificates] Generated API Server key and certificate
[certificates] Generated Service Account signing keys
[certificates] Created keys and certificates in "/etc/kubernetes/pki"
[kubeconfig] Wrote KubeConfig file to disk: "/etc/kubernetes/admin.conf"
[kubeconfig] Wrote KubeConfig file to disk: "/etc/kubernetes/kubelet.conf"
[apiclient] Created API client, waiting for the control plane to become ready
[apiclient] All control plane components are healthy after 7.762803 seconds
[apiclient] Waiting for at least one node to register and become ready
[apiclient] First node is ready after 1.003148 seconds
[apiclient] Creating a test deployment
[apiclient] Test deployment succeeded
[token-discovery] Using token: 8321b6:a535ba541af7623c
[token-discovery] Created the kube-discovery deployment, waiting for it to become ready
[token-discovery] kube-discovery is ready after 1.003423 seconds
[addons] Created essential addon: kube-proxy
[addons] Created essential addon: kube-dns
Your Kubernetes master has initialized successfully!
You should now deploy a pod network to the cluster.
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
    http://kubernetes.io/docs/admin/addons/
You can now join any number of machines by running the following on each node:
kubeadm join --discovery token://8321b6:a535ba541af7623c@10.142.0.6:9898
```

### With invalid token

```
$ sudo ./kubeadm init --discovery token://12345:12345
[kubeadm] WARNING: kubeadm is in alpha, please do not use it for production clusters.
[preflight] Running pre-flight checks
[init] Using Kubernetes version: v1.5.2
[token-discovery] A token has been provided, validating [&{ID:12345 Secret:12345 Addresses:[]}]
token ["12345:12345"] was not of form ["^([a-z0-9]{6})\\:([a-z0-9]{16})$"]
```

### With valid token

```
$ sudo ./kubeadm ex token generate
cd540e:c0e0318e2f4a63b1

$ sudo ./kubeadm init --discovery token://cd540e:c0e0318e2f4a63b1
[kubeadm] WARNING: kubeadm is in alpha, please do not use it for production clusters.
[preflight] Running pre-flight checks
[init] Using Kubernetes version: v1.5.2
[token-discovery] A token has been provided, validating [&{ID:cd540e Secret:c0e0318e2f4a63b1 Addresses:[]}]
[certificates] Generated Certificate Authority key and certificate.
[certificates] Generated API Server key and certificate
[certificates] Generated Service Account signing keys
[certificates] Created keys and certificates in "/etc/kubernetes/pki"
[kubeconfig] Wrote KubeConfig file to disk: "/etc/kubernetes/admin.conf"
[kubeconfig] Wrote KubeConfig file to disk: "/etc/kubernetes/kubelet.conf"
[apiclient] Created API client, waiting for the control plane to become ready
[apiclient] All control plane components are healthy after 13.513305 seconds
[apiclient] Waiting for at least one node to register and become ready
[apiclient] First node is ready after 0.502656 seconds
[apiclient] Creating a test deployment
[apiclient] Test deployment succeeded
[token-discovery] Using token: cd540e:c0e0318e2f4a63b1
[token-discovery] Created the kube-discovery deployment, waiting for it to become ready
[token-discovery] kube-discovery is ready after 2.002457 seconds
[addons] Created essential addon: kube-proxy
[addons] Created essential addon: kube-dns
Your Kubernetes master has initialized successfully!
You should now deploy a pod network to the cluster.
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
    http://kubernetes.io/docs/admin/addons/
You can now join any number of machines by running the following on each node:
kubeadm join --discovery token://cd540e:c0e0318e2f4a63b1@10.142.0.6:9898
```

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -200,17 +200,18 @@ func (i *Init) Validate() error {
...
-	}
-
-	// Exception:
+	// Validate token if any, otherwise generate
 	if i.cfg.Discovery.Token != nil {
-		// Validate token
-		if valid, err := kubeadmutil.ValidateToken(i.cfg.Discovery.Token); valid == false {
-			return err
+		if i.cfg.Discovery.Token.ID != "" && i.cfg.Discovery.Token.Secret != "" {
+			fmt.Printf("[token-discovery] A token has been provided, validating [%+v]\n", i.cfg.Discovery.Token)
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: please just print the token out here in it's string representation, not the go struct.

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -200,17 +200,18 @@ func (i *Init) Validate() error {
...
+			if valid, err := kubeadmutil.ValidateToken(i.cfg.Discovery.Token); valid == false {
+				return err
+			}
+		} else {
+			fmt.Printf("[token-discovery] A token has not been provided, generating one\n")
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: fmt.Println is just fine here

**Review by luxas — COMMENTED** [review]
Great, can you add "integration" tests for this in `cmd/kubeadm/test` that runs kubeadm three times and verifies no token and a valid token exits with 0 and an invalid token exits with a non-zero code?

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Agreed.

File: cmd/kubeadm/app/cmd/init.go (hunk 2 above)
Comment (pires) [inline cmd/kubeadm/app/cmd/init.go]: Agreed.

**Comment by pires** [comment]
@luxas, I'll work on it.

**Comment by luxas** [comment]
Cool, while you're at it, could you split out `cmd/kubeadm/app/util/tokens.go` to a separate `cmd/kubeadm/app/util/token`? 

I guess it's just a matter of a copy of the file, you could do it in a separate commit after the first thing is done or just in another PR

**Comment by pires** [comment]
>  could you split out cmd/kubeadm/app/util/tokens.go to a separate cmd/kubeadm/app/util/token?

@luxas it's wise to do it in a different PR and will do it as soon as this is merged.

**Comment by pires** [comment]
@luxas need your help to get the integration tests working.

**Comment by pires** [comment]
@luxas can I remove the last commit and have it done in another PR by @dmmcquay? `cmd/kubeadm/test` tests were not running at all and this is something Derek has experience with.

**Comment by luxas** [comment]
LGTM

**Pull request merged**